client/finality-grandpa/src/communication/gossip.rs

```diff
@@ -539,21 +542,26 @@ impl<N: Ord> Peers<N> {
 	}
 
 	fn authorities(&self) -> usize {
-		self.inner.iter().filter(|(_, info)| info.roles.is_authority()).count()
+		self.inner.iter().filter(|(_, info)| matches!(info.roles, ObservedRole::Authority)).count()
 	}
 
 	fn non_authorities(&self) -> usize {
-		self.inner.iter().filter(|(_, info)| !info.roles.is_authority()).count()
+		self.inner
+			.iter()
+			.filter(|(_, info)| matches!(info.roles, ObservedRole::Full | ObservedRole::Light))
+			.count()
 	}
 
 	fn reshuffle(&mut self) {
 		let mut lucky_peers: Vec<_> = self.inner
 			.iter()
-			.filter_map(|(id, info)| if !info.roles.is_authority() { Some(id.clone()) } else { None })
+			.filter_map(|(id, info)|
+				if matches!(info.roles, ObservedRole::Full | ObservedRole::Light) { Some(id.clone()) } else { None })
```

Comment:
How about one central `is_authority` function instead of this logic being repeated?


```diff
diff --git a/client/network/src/protocol/event.rs b/client/network/src/protocol/event.rs                                                      │
index ee549fcce..eedcb150d 100644                                                                                                             │
--- a/client/network/src/protocol/event.rs                                                                                                    │
+++ b/client/network/src/protocol/event.rs                                                                                                    │
@@ -97,4 +97,8 @@ impl ObservedRole {                                                                                                         │
        pub fn is_light(&self) -> bool {                                                                                                      │
                matches!(self, ObservedRole::Light)                                                                                           │
        }                                                                                                                                     │
+                                                                                                                                             │
+       pub fn is_authority(&self) -> bool {                                                                                                  │
+               matches!(info.roles, ObservedRole::Full | ObservedRole::Light | ObservedRole::OurSentry)                                      │
+       }                                                                                                                                     │
 }  
```

Reply:
Because this is ambiguous. `OurSentry` and `OurGuardedValidator` are neither authority nor non-authority. They are treated as a third category for which we always broadcast.